framework/src/main/java/org/checkerframework/common/wholeprograminference/WholeProgramInferenceScenes.java

```diff
@@ -452,6 +460,44 @@ public void updateInferredMethodReturnType(
         AnnotatedTypeMirror rhsATM = atf.getAnnotatedType(retNode.getTree().getExpression());
         helper.updateAnnotationSetInScene(
                 method.returnType, atf, jaifPath, rhsATM, lhsATM, TypeUseLocation.RETURN);
+
+        // Now, update overridden methods based on the implementation we just saw.
```

Comment:
Performing this for each return statement is inefficient.  It would be enough to propagate the information once at the end of WPI, for the final return type.
In addition, it's inefficient to propagate (elsewhere) when the overriding relationship is discovered.  More information may be discovered but not propagated.  This could lead to an extra iteration of WPI, which would be very costly.

It would be better to record overriding relationships as they are discovered, then do all propagation related to them at the end of WPI.

Reply:
This seems like premature optimization to me. WPI is already inefficient. I will add a comment with your concern about performance, so that if performance becomes a problem for WPI we can revisit it.